smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.typescript.codegen.integration;
+
+import java.util.List;
+import java.util.Set;
+import software.amazon.smithy.model.knowledge.HttpBinding;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Abstract implementation useful for all protocols that have HTTP document bodies.
+ *
+ * This is distinct from the {@code DocumentShapeSerdeVisitor} implementations in
+ * that it handles serde of documents and delegates generation of serde for shapes
+ * that make up those documents.
+ */
+public abstract class HttpDocumentBodyGenerator implements HttpValueProvider {
+
+    /**
+     * Gets the default content-type when a document is synthesized in the body.
+     *
+     * @return Returns the default content-type.
+     */
+    protected abstract String getDocumentContentType();
+
+    /**
+     * Gets the visitor used to generate serialization functions for shapes.
+     *
+     * @param context The generation context.
+     * @return Returns the serialization visitor implementation.
+     */
+    protected abstract DocumentShapeSerializationVisitor getDocumentShapeSerializer(GenerationContext context);
+
+    /**
+     * Gets the visitor used to generate deserialization functions for shapes.
+     *
+     * @param context The generation context.
+     * @return Returns the deserialization visitor implementation.
+     */
+    protected abstract DocumentShapeDeserializationVisitor getDocumentShapeDeserializer(GenerationContext context);
+
+    /**
+     *
```

Comment:
Missing docs.

Why is this necessary if there's already generateSharedComponents?